src/components/application_manager/src/application_manager_impl.cc

```diff
@@ -4725,42 +4726,70 @@ bool ApplicationManagerImpl::IsAppSubscribedForWayPoints(
   return IsAppSubscribedForWayPoints(app.app_id());
 }
 
-void ApplicationManagerImpl::SubscribeAppForWayPoints(uint32_t app_id) {
+void ApplicationManagerImpl::SubscribeAppForWayPoints(uint32_t app_id,
+                                                      bool response_from_hmi) {
   SDL_LOG_AUTO_TRACE();
   sync_primitives::AutoLock lock(subscribed_way_points_apps_lock_);
   SDL_LOG_DEBUG("Subscribing " << app_id);
   subscribed_way_points_apps_list_.insert(app_id);
+  if (response_from_hmi) {
+    subscribed_to_hmi_way_points_ = true;
+  }
   SDL_LOG_DEBUG("There are applications subscribed: "
                 << subscribed_way_points_apps_list_.size());
-  if (way_points_data_) {
-    smart_objects::SmartObjectSPtr way_point_notification_ =
-        std::make_shared<smart_objects::SmartObject>(*way_points_data_);
-    (*way_point_notification_)[strings::params][strings::connection_key] =
+  if (GetAppServiceManager().FindWayPointsHandler() != nullptr) {
+    auto service = GetAppServiceManager().ActiveServiceForType(
+        EnumToString(mobile_apis::AppServiceType::NAVIGATION));
+    auto it = mobile_way_points_data_.find(service->connection_key);
+    if (mobile_way_points_data_.end() == it) {
+      SDL_LOG_DEBUG("No waypoint data provided by app service provider yet");
+      return;
+    }
+    smart_objects::SmartObjectSPtr way_point_notification =
+        std::make_shared<smart_objects::SmartObject>(it->second);
+    (*way_point_notification)[strings::params][strings::connection_key] =
         app_id;
-    GetRPCService().SendMessageToMobile(way_point_notification_);
+    GetRPCService().SendMessageToMobile(way_point_notification);
```

Comment:
Add return message here, or else mobile could be send two notifications?

Reply:
Ah, should just be in an else-if

Reply:
Fixed in https://github.com/smartdevicelink/sdl_core/pull/3573/commits/109813ab47dcd266f0e66e66450f20eda4b6057b